Implement Metadata filtering on Pgvector with 3 implementations JSON, JSONB and Columns

humcqc opened the pull request:
Implement Metadata filtering on Pgvector with 3 implementations JSON, JSONB Columns
@langchain4j, @jmartisk,  @sebastienblanc  as discuss in quarkus-langchain4j

Example: https://github.com/langchain4j/langchain4j-examples/blob/main/pgvector-example/src/main/java/PgVectorEmbeddingStoreWithMetadataExample.java

Related: https://github.com/quarkiverse/quarkus-langchain4j/pull/410

humcqc committed:
Implement Metadata filtering on Pgvector with 3 implementations JSON, JSONB and Columns

humcqc commented:
Still in progress, I will add Datasource Management in PgVectorEmbeddingStore to simplify the integration

humcqc commented:
@langchain4j, @jmartisk, @sebastienblanc should be ok to review. Thanks
@langchain4j is there a plan to move to java 17 ?

sebastienblanc commented:
@humcqc Looks nice ! I'm just wondering how easy it will be to integrate with the current quarkus extension as it's using lombok and GSON but that's just my no partial opinion 😀

humcqc commented:
I think it should be fine I'm using the same code in my quarkus app. 
GSON is already part of quarkus bom. 
For Lombok I will add the dependencies in quarkiverse-langchain and see if it's ok.

sebastienblanc reviewed (approved):
LGTM

humcqc commented:
The integration is ok and simpler , we just have now to create the DataSourcePgVectorEmbeddingStore in the recorder with quarkus datasource and quarkus config.

I still need to do some test , it seems there is an issue with the datasource when running in quarkus:
With simple PGSimpleDataSource it's working fine, with agroal datasource: java.lang.RuntimeException: java.sql.SQLException: Sorry, acquisition timeout!

humcqc commented:
Should be ok, just waiting for an answer to my renaming question to simplify.

humcqc committed:
- add default constructor for bean validation

jmartisk commented:
Great work! Have you already got an updated version of the Quarkus integration that is based on this? It would be much more efficient to review both together

humcqc commented:
ok will do.
I would like first to rename DataSourcePgVectorEmbeddingStore to PgVectorEmbeddingStore if possible. And impact the example to use a datasource instead of database parameters. What do you think ?

jmartisk commented:
Could we potentially merge them into one class that has two constructors, one that takes connection parameters, and one that takes a datasource? I think both approaches have their merits.

humcqc committed:
- Use one Store with 2 builders

humcqc commented:
Done :)

humcqc commented:
https://github.com/quarkiverse/quarkus-langchain4j/pull/410 updated with this version

jmartisk commented on langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/PgVectorEmbeddingStore.java:
```diff
@@ -250,15 +307,16 @@ private void addAllInternal(
+                    table, metadataHandler.columnsNames(),
+                    join(",", nCopies(metadataHandler.nbMetadataColumns(), "?")),
+                    metadataHandler.insertClause());
             PreparedStatement upsertStmt = connection.prepareStatement(query);
```
Should probably use a try-with-resources block, this statement doesn't get closed when an exception occurs

humcqc committed:
- Fix some leaks in case of errors and add extension checks in getConnection as initially

jmartisk commented on langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/ColumnsMetadataHandler.java:
```diff
@@ -0,0 +1,111 @@
+        String indexTypeSql = indexType == null ? "" : "USING " + indexType;
+        this.indexes.stream().map(String::trim)
+                .forEach(index -> {
+                    String indexSql = String.format("create index %s_%s on %s %s ( %s )",
```
shouldn't there be `IF NOT EXISTS`?

jmartisk commented on langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/JSONBMetadataHandler.java:
```diff
@@ -0,0 +1,52 @@
+            String index = str.trim();
+            String indexName = formatIndex(index);
+            try {
+                String indexSql = String.format("create index %s_%s on %s %s ( %s )",
```
shouldn't there be `IF NOT EXISTS`?

jmartisk commented on langchain4j-pgvector/src/main/java/dev/langchain4j/store/embedding/pgvector/JSONMetadataHandler.java:
```diff
@@ -0,0 +1,92 @@
+ * This class handle JSON and JSONB Filter mapping
+ */
+public class JSONMetadataHandler implements MetadataHandler {
+    private static final Gson GSON = new Gson();
```
Please switch to using `dev.langchain4j.internal.JsonCodec` instead of GSON directly. This is an abstraction over different JSON implementations and we need it to be able to avoid dragging the dependency on GSON into Quarkus.

humcqc committed:
- @jmartisk review 1

humcqc committed:
- Fix java 8 compilation

jmartisk reviewed (approved):
Awesome work! From my PoV, this is ready.

@langchain4j, please have a look too, I tested it together with the Quarkus integration at https://github.com/quarkiverse/quarkus-langchain4j/pull/410 and it works for me.

humcqc committed:
@langchain4j review 1:
- Use MetadataStorageMode enum instead of type string
- ensure datasource not null
- change default PgVectorEmbeddingStore builder to avoid builder breaking change
- use utils getOrDefault
- remove unused method createMetadataExtensions
- remove default createMetadataIndexes
- simplify IT
- rename MetadataConfig
- Change columnNames to list and remove nbMetadataColumns
- Create metadata index just if createTable
- make all new classes package-private, where possible

WIP, waiting for answers to finalize.

humcqc commented:
Thanks @langchain4j for the precise and great review. I've done 90% of it, waiting for some precision for columnDefinition, i would like to keep things flexible to use sql definition

humcqc committed:
@langchain4j review 2:
- Fix indexes property

WIP, waiting for answers to finalize.

humcqc committed:
@langchain4j review 2:
- Fix unclear javadoc

WIP, waiting for answers to finalize.

humcqc committed:
@langchain4j review 3:
- Add MetadataColumDefinition

WIP, waiting for answers to finalize.

humcqc commented:
weird , IT should have a comment, I modify the the column definition in this commit : https://github.com/langchain4j/langchain4j/pull/851#:~:text=%40langchain4j-,review%203%3A,-%E2%80%A6
I use a specific class to handle Column Definitions, but i keep the flexibility for types and other SQL attributes.

humcqc committed:
@langchain4j review 4: Add createMetadataIndexes unit tests

humcqc committed:
@langchain4j review 5: Add a test to verify that store created in version 0.29 still work in 0.30+

humcqc commented:
@langchain4j done.

dliubarskyi committed:
Merge branch 'main' into main

humcqc commented:
thanks @langchain4j , example added: https://github.com/langchain4j/langchain4j-examples/pull/80

humcqc commented:
Hello, I will have a look.

Le jeu. 7 nov. 2024, 09:49, LangChain4j ***@***.***> a écrit :

> Hi @humcqc <https://github.com/humcqc>, could you please check
> https://app.codacy.com/gh/langchain4j/langchain4j/issues/131242390118 and
> https://app.codacy.com/gh/langchain4j/langchain4j/issues/131242390121 ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/langchain4j/langchain4j/pull/851#issuecomment-2461650915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A477YIXOPPXFWF3AB5FDZILZ7MSSHAVCNFSM6AAAAABRKVJRDWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRRGY2TAOJRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
